geode-core/src/upgradeTest/java/org/apache/geode/security/AuthExpirationFunctionDUnitTest.java

```diff
@@ -65,77 +76,327 @@
+  private MemberVM serverVM2;
 
   @Rule
-  public ClusterStartupRule lsRule = new ClusterStartupRule();
+  public ClusterStartupRule lsRule = new ClusterStartupRule(4);
+
+  @Rule
+  public ClientCacheRule clientCacheRule = new ClientCacheRule();
 
   @Before
-  public void setup() throws Exception {
-    Properties properties = new Properties();
-    properties.setProperty(SECURITY_MANAGER, ExpirableSecurityManager.class.getName());
-    properties.setProperty(ConfigurationProperties.SERIALIZABLE_OBJECT_FILTER,
+  public void setup() {
+    MemberVM locatorVM =
+        lsRule.startLocatorVM(0, l -> l.withSecurityManager(ExpirableSecurityManager.class));
+    int locatorPort = locatorVM.getPort();
+
+    Properties serverProperties = new Properties();
+    serverProperties.setProperty(SECURITY_MANAGER, ExpirableSecurityManager.class.getName());
+    serverProperties.setProperty(ConfigurationProperties.SERIALIZABLE_OBJECT_FILTER,
         "org.apache.geode.management.internal.security.TestFunctions*");
-    serverVM = lsRule.startServerVM(0, properties);
+    serverProperties.setProperty(GROUPS, "group");
+    serverProperties.setProperty(USER_NAME, "test");
+    serverProperties.setProperty(PASSWORD, "test");
 
-    serverVM.invoke(() -> {
+    serverVM0 = lsRule.startServerVM(1, serverProperties, locatorPort);
+    serverVM1 = lsRule.startServerVM(2, serverProperties, locatorPort);
+    serverVM2 = lsRule.startServerVM(3, serverProperties, locatorPort);
+
+    VMProvider.invokeInEveryMember(() -> {
       Objects.requireNonNull(ClusterStartupRule.getCache())
           .createRegionFactory(RegionShortcut.REPLICATE).create("region");
-    });
-    int serverPort = serverVM.getPort();
-    clientVM = lsRule.startClientVM(1, clientVersion, c1 -> c1
+      Objects.requireNonNull(ClusterStartupRule.getCache())
+          .createRegionFactory(RegionShortcut.PARTITION).create("partitionRegion");
+    }, serverVM0, serverVM1, serverVM2);
+
+    VMProvider.invokeInEveryMember(() -> writeFunction = new TestFunctions.WriteFunction(),
+        serverVM0, serverVM1, serverVM2);
+
+    clientCacheRule
         .withProperty(SECURITY_CLIENT_AUTH_INIT, UpdatableUserAuthInitialize.class.getName())
         .withPoolSubscription(true)
-        .withServerConnection(serverPort));
+        .withLocatorConnection(locatorPort);
+  }
 
-    VMProvider.invokeInEveryMember(() -> writeFunction = new TestFunctions.WriteFunction(),
-        serverVM, clientVM);
+  @Test
+  public void clientShouldReAuthenticateWhenCredentialExpiredAndFunctionExecutionOnServerSucceed()
+      throws Exception {
+    ClientCache clientCache = clientCacheRule.createCache();
+    UpdatableUserAuthInitialize.setUser("data1");
+    writeFunction = new TestFunctions.WriteFunction();
+
+    ResultCollector rc = onServer(clientCache.getDefaultPool()).execute(writeFunction);
+    assertThat(((ArrayList) rc.getResult()).get(0))
+        .isEqualTo(TestFunctions.WriteFunction.SUCCESS_OUTPUT);
```

Comment:
You should always try to cast to an interface rather than a concrete implementation:
```
List<?> result = (List) rc.getResult();
assertThat(result.get(0))
    .isEqualTo(TestFunctions.WriteFunction.SUCCESS_OUTPUT);
```